cluster-controller/src/main/java/io/strimzi/controller/cluster/operations/cluster/KafkaClusterOperations.java

```diff
@@ -230,13 +294,35 @@ public void create(String namespace, String name, Handler<AsyncResult<Void>> han
         }
     };
 
+    private final CompositeOperation<TopicController> deleteTopicController = new CompositeOperation<TopicController>() {
+
+        @Override
+        public Future<?> composite(String namespace, ClusterOperation<TopicController> clusterOp) {
+            TopicController topicController = clusterOp.cluster();
+            Future<Void> fut = deploymentOperations.delete(namespace, topicController.getName());
+            return fut;
+        }
+
+        @Override
+        public ClusterOperation<TopicController> getCluster(String namespace, String name) {
+            Deployment dep = deploymentOperations.get(namespace, TopicController.topicControllerName(name));
+            return new ClusterOperation<>(TopicController.fromDeployment(namespace, name, dep), null);
+        }
+    };
+
     @Override
     protected void delete(String namespace, String name, Handler<AsyncResult<Void>> handler) {
-        execute(CLUSTER_TYPE_KAFKA, OP_DELETE, namespace, name, deleteKafka, ar -> {
-            if (ar.failed()) {
-                handler.handle(ar);
+        execute(CLUSTER_TYPE_KAFKA, OP_DELETE, namespace, name, deleteKafka, kafkaDone -> {
+            if (kafkaDone.failed()) {
+                handler.handle(kafkaDone);
             } else {
-                execute(CLUSTER_TYPE_ZOOKEEPER, OP_DELETE, namespace, name, deleteZk, handler);
+                execute(CLUSTER_TYPE_ZOOKEEPER, OP_DELETE, namespace, name, deleteZk, zookeeperDone -> {
+                    if (zookeeperDone.failed()) {
+                        handler.handle(zookeeperDone);
+                    } else {
+                        execute(CLUSTER_TYPE_TOPIC_CONTROLLER, OP_DELETE, namespace, name, deleteTopicController, handler);
```

Comment:
Eh, I think the TC should be deleted as first, not as last. At least in ideal case we should try to delete it before we delete Kafka and ZK where it is connected.

Reply:
Agree it should be the opposite order than creation.